### PR TITLE
Soccer API (MVP) #2

### PR DESCRIPTION
**Pull request by nghiapham1026**
- Get all matches results from 1 day prior to today at /espn/getPastResults:
- Return and save the name of teams
- Return and save dates of games
- Return and save the logos of each team
- Return and save the score of matches
- Post match data into database at the "sampleMatchData" collection through the postData() function
- Define a new data schema for storage in the database. This structure contains an array of matches with their properties
266165521-3ed71dd2-d92e-46fa-be33-2a0e379fd6e7
266165776-25f54996-ec88-4d3c-9119-b2154fee807c
Screenshot 2023-09-07 071759

### CODE REVIEW

File: controllers/todo.js [hunk 1]
```diff
@@ -1,27 +1,34 @@
...
     })
     .catch((err) => {
       res.status(500).json({
-        message: "Fail to create todo!",
+        message: "Failed to add match data",
+        error: err,
       });
     });
-};
+};
```
Comment (saravpereira) [inline controllers/todo.js]: do not reuse todo, create your own controller with the correct variable and file name

File: controllers/webscrapeController.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 const puppeteer = require("puppeteer");
 const { espnScoreUrl, defaultLogo } = require("./constants");
 
-async function scrapeEspn(startDate, endDate, res) {
+async function scrapeEspn(startDate, endDate, req, res) {
```
Comment (saravpereira) [inline controllers/webscrapeController.js]: req is not used here, why are you adding it back to the argument?

File: index.js [hunk 3]
```diff
@@ -26,12 +30,31 @@ app.get("/espn/getGamesByDates", (req, res) => {
 
   try {
     const { startDate, endDate } = getDateRange(userEndDate);
-    scrapeController.scrapeEspn(startDate, endDate, res);
+    scrapeController.scrapeEspn(startDate, endDate, req, res);
```
Comment (saravpereira) [inline index.js]: remove req if not used

File: models/todo.js [hunk 4]
```diff
@@ -1,19 +1,49 @@
...
-  date: {
-    type: String,
-    required: true,
-  },
-});
+  { collection: "sampleMatchData"}
+);
 
-module.exports = mongoose.model("Todo", todoSchema);
+module.exports = mongoose.model("Todo", todoSchema);
```
Comment (saravpereira) [inline models/todo.js]: do not reuse todo, create your own models with the correct variable and file name

File: utils/postData.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const axios = require('axios');
+const config = require('../controllers/constants');
+
+async function postData(allMatchData) {
+  try {
+    const endpoint = `${config.serverUrl}/api/todo`;
+    const response = await axios.post(endpoint, { matches: allMatchData });
+    console.log(`Data sent to ${endpoint}:`, response.data);
```
Comment (saravpereira) [inline utils/postData.js]: api calls are not util functions. place this in the correct location

File: controllers/constants.js [hunk 6]
```diff
@@ -1,5 +1,11 @@
 const espnUrl = "https://www.espn.com";
 const espnScoreUrl = `${espnUrl}/soccer/scoreboard`;
 const defaultLogo = "https://www.seekpng.com/png/full/28-289657_espn-soccer-team-logo-default.png";
+const serverUrl = 'http://localhost:8080';
```
Comment (saravpereira) [inline controllers/constants.js]: server url should not be placed in const, add it as a variable in .env file

File: utils/postData.js (hunk 5 above)
Comment (saravpereira) [reply utils/postData.js]: do not call another api responsible for handling the db entry. handle it in the api you are creating

File: controllers/constants.js (hunk 6 above)
Comment (saravpereira) [reply controllers/constants.js]: based on my comment below, this serverUrl isnt even needed

File: routes/postData.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+const Todo = require('../models/match');
```
Comment (saravpereira) [inline routes/postData.js]: this is not a todo list

File: routes/postData.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  } catch (error) {
+    console.error('Error saving data to database:', error);
+    return false;
+  }
+}
+
+module.exports = {
+  postData
+};
```
Comment (saravpereira) [inline routes/postData.js]: this is a controller, not a route.
Move this file to controller file

File: models/todo.js
Comment (saravpereira) [inline models/todo.js:1]: these todo stuff was here as sample reference, do not delete. please bring back all the todo changes you removed

**Review by saravpereira — CHANGES_REQUESTED** [review]
Architecturally, you may need to look into API routes. It use to be there for todo but you removed it all. look at for example: 
https://javascript.plainenglish.io/nodejs-backend-in-2022-setting-up-server-in-5-simple-steps-19074ab861f1

**Comment by nghiapham1026** [comment]
> Architecturally, you may need to look into API routes. It use to be there for todo but you removed it all. look at for example: https://javascript.plainenglish.io/nodejs-backend-in-2022-setting-up-server-in-5-simple-steps-19074ab861f1

The link is behind a paywall by the way.

File: controllers/match.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        message: "Failed to add match data",
+        error: err,
+      });
+    });
+};
+
+exports.getPastResults = async (_, res) => {
+  try {
+    const allMatchData = await scrapeController.scrapeEspn(yesterdayDate, yesterdayDate);
+    const postSuccess = await postData(allMatchData);
```
Comment (saravpereira) [inline controllers/match.js]: put whatever you wrote in postData here instead of separating it into different files

File: models/match.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+          required: true,
+        },
+        awayLogo: {
+          type: String,
+          required: true,
+        },
+      },
+    ],
+  },
+  { collection: "sampleMatchData"}
```
Comment (saravpereira) [inline models/match.js]: sample match data is not a meaningful name. Rename this to matchData

Comment (saravpereira) [reply models/match.js]: when you change the name, delete sampleMatchData from the db.
**do not delete anything else**

File: models/match.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+          required: true,
+        },
+        awayLogo: {
+          type: String,
+          required: true,
+        },
+      },
+    ],
+  },
+  { collection: "MatchData" }
```
Comment (saravpereira) [inline models/match.js]: camelCase please

**Review by saravpereira — APPROVED** [review]

**Pull request merged**